### PR TITLE
Issue #49: usability enhancements

### PR DESCRIPTION
**Pull request by phijor**
Changes:

* list available events (predefined and kernel PMU) with `--list-events`
* list available tracepoint events with `--list-tracepoints`
* show descriptive error when using `x86_adapt`-related switches but `lo2s` was built without support for `x86_adapt`.
* group options in usage section
* add `lo2s [OPTIONS] -a` as invocation pattern in usage section
* give names to arguments to options
* add usage section explaining arguments to options.

### CODE REVIEW

**Comment by bmario** [comment]
Looks good, but I'd still like to improve a few things:

- [ ] `--list-events=tracepoint` should actually be `--list-tracepoints`
- [ ] There should also be a  `--list-knobs`
- [ ] the argument for `-x [ --x86-adapt-cpu-knob ] arg` should be KNOB
- [ ] `--command` shouldn't be listed as option

File: src/config.cpp [hunk 1]
```diff
@@ -103,99 +189,170 @@ const Config& config()
...
+                ->default_value("monotonic-raw"),
+            "Clock used for perf timestamps (see --list-clockids for supported arguments).")
+        ("list-clockids",
+            po::bool_switch(&list_clockids)
+                ->default_value(false),
+            "List all available clockids.")
+        ("list-events",
+            po::value(&list_event_category)
+                ->value_name("CATEGORY")
+                ->implicit_value("all"),
```
Comment (bmario) [inline src/config.cpp]: The help message generated from this looks different than other options.
Considering that I'd like to split this anyways, just turn this into a bool_switch and print predefined and pmu events.
`--list-tracepoints` than should print the list of tracepoints.

File: src/config.cpp [hunk 2]
```diff
@@ -103,99 +189,170 @@ const Config& config()
...
+            po::bool_switch(&config.enable_cct),
+            "Enable call-graph recording.")
+        ("no-ip,n",
+            po::bool_switch(&config.suppress_ip),
+            "Do not record instruction pointers [NOT CURRENTLY SUPPORTED]")
+        ("pid,p",
+            po::value(&config.pid)
+                ->value_name("PID")
+                ->default_value(-1),
+            "Attach to process of given PID.")
```
Comment (bmario) [inline src/config.cpp]: Can we hide the `(=-1)` in the output? Probably something like remove the `default_value()` and add something like this: 

```
if vm.count("pid"):
  config.pid = vm.pid
else:
  config.pid = -1
```

File: src/config.cpp [hunk 3]
```diff
@@ -103,99 +189,170 @@ const Config& config()
...
+                ->default_value(11010113),
+            "Sampling period (in number of events specified by -e).")
+        ("event,e",
+            po::value(&config.sampling_event)
+                ->value_name("EVENT")
+                ->default_value("instructions"),
+            "Interrupt source event for sampling.")
+        ("call-graph,g",
+            po::bool_switch(&config.enable_cct),
+            "Enable call-graph recording.")
```
Comment (bmario) [inline src/config.cpp]: Change help to:
`Record call stack of samples`

File: src/config.cpp [hunk 4]
```diff
@@ -86,12 +148,36 @@ static inline void print_version(std::ostream& os)
...
+                usually requires read/execute permissions on
+                    /sys/kernel/debug,
+                which by default is owned by root:root.
+)";
+
     // clang-format off
     os << "Usage:\n"
           "  " << name << " [options] ./a.out\n"
           "  " << name << " [options] -- ./a.out --option-to-a-out\n"
           "  " << name << " [options] --pid $(pidof some-process)\n"
```
Comment (bmario) [inline src/config.cpp]: Please also add a line for `-a` mode

File: src/config.cpp [hunk 5]
```diff
@@ -86,12 +128,39 @@ static inline void print_version(std::ostream& os)
...
+    TRACEPOINT  Name of a kernel tracepoint event.  Format is either
+                    <group>:<name>
+                or
+                    <group>/<name>.
+                Tracepoint events can be found at
+                    /sys/kernel/debug/tracing/events/<group>/<name>.
+                Accessing tracepoints (even for --list-tracepoints)
+                usually requires read/execute permissions on
+                    /sys/kernel/debug
+)";
```
Comment (bmario) [inline src/config.cpp]: This should also contain an entry for `KNOB` and `CLOCKID`.
Also, mention the list command for each one.

File: src/config.cpp [hunk 6]
```diff
@@ -240,25 +393,35 @@ void parse_program_options(int argc, const char** argv)
...
-    if (list_clockids || list_events)
     {
         if (list_clockids)
         {
-            std::cout << "Available clockids:\n";
-            for (const auto& clock : lo2s::time::ClockProvider::get_descriptions())
-            {
-                std::cout << " * " << clock.name << '\n';
-            }
+            std::cout << io::make_argument_list("available clockids",
```
Comment (bmario) [inline src/config.cpp]: This should more look like:

```
auto clock_ids = time::ClockProvider::get_descriptions();
std::cout << io::make_argument_list("available clockids", std::begin(clock_ids), std::end(clock_ids));
```

File: include/lo2s/io.hpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                                         InputIterator last)
+{
+    return ArgumentList<InputIterator>{ description, first, last };
+}
+
+/**
+ * \brief Creates ArgumentList object from a fixed-size array of elements
+ */
+template <class T, std::size_t N>
+inline constexpr auto make_argument_list(const char* description, const T (&array)[N])
```
Comment (bmario) [inline include/lo2s/io.hpp]: This overload is not necessary, use `std::begin()` and `std::end()` instead. See below.

File: include/lo2s/io.hpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+struct ArgumentList
+{
+    ArgumentList() = delete;
+
+    /**
+     * \param description   description of list contents printed before list
+     * \param first         iterator pointing to first item to print
+     * \param last          iterator pointing past the last item to print
+     * */
+    constexpr ArgumentList(const char* description, InputIterator first, InputIterator last)
```
Comment (bmario) [inline include/lo2s/io.hpp]: > `const char*`

We are not in C, please use `const std::string&` instead.

File: src/config.cpp [hunk 9]
```diff
@@ -73,6 +80,41 @@ void validate(boost::any& v, const std::vector<std::string>&, SwitchCounter*, lo
     }
 }
 
+static inline void list_arguments_sorted(std::ostream& os, const char* description,
```
Comment (bmario) [inline src/config.cpp]: const std::string&

File: src/perf/tracepoint/format.cpp [hunk 10]
```diff
@@ -98,6 +98,29 @@ void EventFormat::parse_format_line(const std::string& line)
     }
 }
 
+std::vector<std::string> EventFormat::get_tracepoint_event_names() try
```
Comment (bmario) [inline src/perf/tracepoint/format.cpp]: You're missing a pair of curly braces.

File: src/config.cpp [hunk 11]
```diff
@@ -86,12 +128,39 @@ static inline void print_version(std::ostream& os)
 static inline void print_usage(std::ostream& os, const char* name,
                                const po::options_description& desc)
 {
+    static constexpr char argument_detail[] =
+        R"(
+Arguments to options:
+    PATH        A filesystem path.  It can be either relative to the
```
Comment (bmario) [inline src/config.cpp]: Please include the changes of ae40d227de88d221d3151bf63ed218c959d04f4f on master

File: src/perf/tracepoint/format.cpp (hunk 10 above)
Comment (phijor) [reply src/perf/tracepoint/format.cpp]: Today I learned that this is actually a valid syntax construct: a [function-try-block](http://en.cppreference.com/w/cpp/language/function-try-block). Will change nonetheless.

**Pull request merged**